src/Catalyst.Core.Modules.Kvm.Tests/IntegrationTests/CatalystVirtualMachineTests.cs

```diff
@@ -188,6 +194,70 @@ public void Can_Invoke_Chain_Id()
+            trace.Entries.Last().Memory.First().Should().Be("378d0caaaa3855f1b38693c1d6ef004fd118691c95c959d4efa950d6d6fcf7c1");
+        }
+        
+        [Fact]
+        [Trait(Traits.TestType, Traits.IntegrationTest)]
+        public void Ed25519_verify_precompile()
+        {
+            HashProvider hashProvider = new HashProvider(HashingAlgorithm.GetAlgorithmMetadata("blake2b-256"));
+            
+            FfiWrapper cryptoContext = new FfiWrapper();
+            var privateKey = cryptoContext.GeneratePrivateKey();
+            
+            var data = Encoding.UTF8.GetBytes("Testing testing 1 2 3");
+            var message = hashProvider.ComputeMultiHash(data).Digest;
+            
+            var signingContext = Encoding.UTF8.GetBytes("Testing testing 1 2 3 context");
+            var context = hashProvider.ComputeMultiHash(signingContext).Digest;
+            
+            var signature = cryptoContext.Sign(privateKey, message, context);
+            var signatureBytes = signature.SignatureBytes;
+            
+            var publicKeyBytes = privateKey.GetPublicKey().Bytes;
+            var signatureFromBytes = cryptoContext.GetSignatureFromBytes(signatureBytes,
+                publicKeyBytes);
+
+            cryptoContext.Verify(signatureFromBytes, message, context)
+               .Should().BeTrue("signature generated with private key should verify with corresponding public key");
+
+            var pushToMemoryAt0 = "7f" + message.ToHexString(false) + "600052";
+            var pushToMemoryAt32 = "7f" + signatureBytes.AsSpan(0, 32).ToHexString(false) + "602052";
+            var pushToMemoryAt64 = "7f" + signatureBytes.AsSpan(32, 32).ToHexString(false) + "604052";
+            var pushToMemoryAt96 = "7f" + context.ToHexString(false) + "606052";
+            var pushToMemoryAt128 = "7f" + publicKeyBytes.ToHexString(false) + "608052";
+            
+            Address edAddress = Address.FromNumber(2 + KatVirtualMachine.CatalystPrecompilesAddressingSpace);
+            string addressCode = edAddress.Bytes.ToHexString(false);
+            
+            var code = Bytes.FromHexString("0x" + pushToMemoryAt0 +
+                pushToMemoryAt32 +
+                pushToMemoryAt64 +
+                pushToMemoryAt96 +
+                pushToMemoryAt128 + "6020600060a0600073" + addressCode + "45fa00");
+            _testOutputHelper.WriteLine(code.ToHexString());
+            
+            var txTracer = RunVirtualMachine(code);
+            var serializer = new EthereumJsonSerializer();
+            var trace = txTracer.BuildResult();
+            _testOutputHelper.WriteLine(serializer.Serialize(trace, true));
+            trace.Entries.Last().Stack.First().Should().Be("0000000000000000000000000000000000000000000000000000000000000001");
+            trace.Entries.Last().Memory.First().Should().Be(publicKeyBytes.ToHexString(false));
```

Comment:
Hi @tkstanczak, I created a new test to check I'm understanding the precompile logic. I wanted to test that the verification can fail. The test is very similar to this one but I generate a second public key and add those bytes instead of the correct ones for variable ```pushToMemoryAt128```. Then I changed this last line to ```trace.Entries.Last().Memory.First().Should().Be(Bytes.Empty.ToHexString(false));``` but the test fails when I think it should pass. Have I made a mistake somewhere?

Reply:
it should be trace.Entries.Last().Stack.First().Should().Be(all zeros here)

Reply:
the memory will stay the same

Reply:
I will add a test for failure, but memory is the precompile input, stack has the result of the precompile call

Reply:
@franssl I am free to explain that on a call next week / on slack

Reply:
added a test on failure
the fail logic was wrong however not in the place that you pointed at

memory will return now 0 or 1 as a result of verification
stack contains the call success result (1 is OK - it was possible to run verification which may have returned false or true, while 0 means a call failure - input length was invalid)

Reply:
Yeah thanks for the explanation that makes sense